// @esri/arcgis-rest-developer-credentials/dist/cjs/shared/generateApiKeyToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateApiKeyToken = void 0;
const arcgis_rest_request_1 = require("@esri/arcgis-rest-request");
const getRegisteredAppInfo_js_1 = require("./getRegisteredAppInfo.js");
const arcgis_rest_portal_1 = require("@esri/arcgis-rest-portal");
async function generateApiKeyToken(options) {
    const portal = (0, arcgis_rest_portal_1.getPortalUrl)(options);
    const url = `${portal}/oauth2/token`;
    const appInfo = await (0, getRegisteredAppInfo_js_1.getRegisteredAppInfo)({
        itemId: options.itemId,
        authentication: options.authentication
    });
    const params = {
        client_id: appInfo.client_id,
        client_secret: appInfo.client_secret,
        apiToken: options.apiKey,
        regenerateApiToken: true,
        grant_type: "client_credentials"
    };
    // authentication is not being passed to the request because client_secret acts as the auth
    return (0, arcgis_rest_request_1.request)(url, {
        params
    });
}
exports.generateApiKeyToken = generateApiKeyToken;
